service/history/workflow/mutable_state_impl.go

```diff
@@ -6732,70 +6732,73 @@ func (ms *MutableStateImpl) disablingTransitionHistory() bool {
+		override.GetBehavior() == enumspb.VERSIONING_BEHAVIOR_PINNED {
+		return override.GetDeployment()
 	}
 	return versioningInfo.GetDeployment()
 }
 
-func (ms *MutableStateImpl) GetRedirectInfo() *persistencespb.WorkflowExecutionInfo_VersioningInfo_RedirectInfo {
-	return ms.GetExecutionInfo().GetVersioningInfo().GetRedirectInfo()
+func (ms *MutableStateImpl) GetDeploymentTransition() *workflowpb.DeploymentTransition {
+	return ms.GetExecutionInfo().GetVersioningInfo().GetDeploymentTransition()
 }
 
-// GetVersioningBehavior returns the effective versioning behavior for the workflow.
-func (ms *MutableStateImpl) GetVersioningBehavior() enumspb.VersioningBehavior {
+// GetEffectiveVersioningBehavior returns the effective versioning behavior in the following
+// order:
+//  1. VersioningOverride.Behavior: this is returned when user has set a behavior override
+//     at wf start time, or later via UpdateWorkflowExecutionOptions.
+//  2. Behavior: this is returned when there is no override (most common case). Behavior is
+//     set based on the worker-sent deployment in the latest WFT completion.
+func (ms *MutableStateImpl) GetEffectiveVersioningBehavior() enumspb.VersioningBehavior {
 	versioningInfo := ms.GetExecutionInfo().GetVersioningInfo()
 	if versioningInfo == nil {
 		return enumspb.VERSIONING_BEHAVIOR_UNSPECIFIED
-	} else if redirectInfo := versioningInfo.GetRedirectInfo(); redirectInfo != nil &&
-		redirectInfo.GetBehaviorOverride() != enumspb.VERSIONING_BEHAVIOR_UNSPECIFIED {
-		return redirectInfo.GetBehaviorOverride()
-	} else if versioningInfo.GetBehaviorOverride() != enumspb.VERSIONING_BEHAVIOR_UNSPECIFIED {
-		return versioningInfo.GetBehaviorOverride()
+	} else if override := versioningInfo.GetVersioningOverride(); override != nil {
+		return override.GetBehavior()
 	}
 	return versioningInfo.GetBehavior()
 }
 
-// StartDeploymentRedirect starts a redirect to the given deployment. If the workflow is pinned,
-// the redirect will be rejected unless it's initiated by an override. Returns true if the requested
-// redirect is started. Starting a new redirect replaces possible existing redirect without
-// rescheduling activities.
+// StartDeploymentTransition starts a transition to the given deployment. Returns true
+// if the requested transition is started. Starting a new transition replaces possible
+// existing ongoing transition without rescheduling activities. If the workflow is
+// pinned, the transition won't start.
 // TODO (shahab): validate source deployment
-func (ms *MutableStateImpl) StartDeploymentRedirect(
-	deployment *deploymentpb.Deployment,
-	behaviorOverride enumspb.VersioningBehavior,
-) bool {
-	if deployment.Equal(ms.GetCurrentDeployment()) {
-		// Not a deployment change.
+func (ms *MutableStateImpl) StartDeploymentTransition(deployment *deploymentpb.Deployment) bool {
+	if deployment.Equal(ms.GetEffectiveDeployment()) {
+		// Not an effective deployment change.
```

Comment:
if
```
baseDeployment=A
overrideDeployment=B
transitionDeployment=B
```
I think we still want to do the transition, because we want baseDeployment to be changed to B

Reply:
I think this should be `if deployment.Equal(ms.execution_info.versioning_info.deployment)`

Reply:
but you don't start a transition in unsafe move, right? I think in that case, there would not be a transition, but baseDeployment will be updated anyways when the next wft completes.